# Sorting and cleaning 
---



In order to effectively analyse a dataset, often we need to prepare it first. 
Before a dataset is ready to be analysed we might need to:  

* sort the data (can be a series or dataframe)  
* remove any NaN values or drop NA values   
* remove duplicate records (identical rows)  
* normalise data in dataframe columns so that has a common scale [reference](https://towardsai.net/p/data-science/how-when-and-why-should-you-normalize-standardize-rescale-your-data-3f083def38ff#:~:text=Similarly%2C%20the%20goal%20of%20normalization,dataset%20does%20not%20require%20normalization.&text=So%20we%20normalize%20the%20data,variables%20to%20the%20same%20range.)

## Sorting the data  
---


Typically we want to sort data by the values in one or more columns in the dataframe  

To sort the dataframe by series we use the pandas function **sort_values()**.  

By default `sort_values()` sorts into ascending order.

* sort by a single column e.g.
  * `df.sort_values("Make") `
* sort by multiple columns e.g. 
  * `df.sort_values(by = ["Model", "Make"]) `
    * this sorts by Model, then my Make 
* sort in *descending* order
  * `df.sort_values(by = "Make", ascending = False)`
  * `df.sort_values(by = ["Make", "Model"], ascending = False])`  

Dataframes are mostly immutable, changes like sort_values do not change the dataframe permanently, they just change it for the time that the instruction is being used.

`df.sort_values(by='Make')` *dataframe is now in sorted order and can be copied to a new dataframe*  
`df` *original dataframe, df, will be as it was - unsorted* 

To split the dataframe after sorting, do this in the same instruction, e.g.:

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]]`

This sorts on Make and then Model in descending order, then splits off the Make and Model columns.

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]].head()`

This sorts on Make and then Model, then splits off the Make and Model columns and then splits off the first 5 rows.

### Exercise 1 - get data, sort by happiness score 
---

Read data from the Excel file on Happiness Data at this link: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true

Display first 5 rows of data  

The data is currently sorted by Happiness Rank...
*  sort the data by Happiness Score in ascending order
*  display sorted table

**Test output**:  
The lowest score (displayed first) is 2.839, Togo  
The highest score (displayed last) is 7.587, Switzerland  



In [104]:
import pandas as pd

url = "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true"

happiness_df = pd.read_excel(url)

# Display first 5 rows
display(happiness_df.head())

# Display table sorted by Happiness Score
display(happiness_df.sort_values("Happiness Score"))

Country          Region  Happiness Rank  Happiness Score  \
0  Switzerland  Western Europe               1            7.587   
1      Iceland  Western Europe               2            7.561   
2      Denmark  Western Europe               3            7.527   
3       Norway  Western Europe               4            7.522   
4       Canada   North America               5            7.427   

   Standard Error  Economy (GDP per Capita)   Family  \
0         0.03411                   1.39651  1.34951   
1         0.04884                   1.30232  1.40223   
2         0.03328                   1.32548  1.36058   
3         0.03880                   1.45900  1.33095   
4         0.03553                   1.32629  1.32261   

   Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                   0.94143  0.66557                        0.41978   
1                   0.94784  0.62877                        0.14145   
2                   0.87464  0.64938                        0.48357   
3                   0.88521  0.66973                        0.36503   
4                   0.90563  0.63297                        0.32957   

   Generosity  Dystopia Residual  
0     0.29678            2.51738  
1     0.43630            2.70201  
2     0.34139            2.49204  
3     0.34699            2.46531  
4     0.45811            2.45176

Country                           Region  \
157                      Togo               Sub-Saharan Africa   
156                   Burundi               Sub-Saharan Africa   
155                     Syria  Middle East and Northern Africa   
154                     Benin               Sub-Saharan Africa   
153                    Rwanda               Sub-Saharan Africa   
152               Afghanistan                    Southern Asia   
151              Burkina Faso               Sub-Saharan Africa   
150               Ivory Coast               Sub-Saharan Africa   
149                    Guinea               Sub-Saharan Africa   
148                      Chad               Sub-Saharan Africa   
147  Central African Republic               Sub-Saharan Africa   
146                Madagascar               Sub-Saharan Africa   
145                  Tanzania               Sub-Saharan Africa   
144                  Cambodia                Southeastern Asia   
143                     Niger               Sub-Saharan Africa   
142                     Gabon               Sub-Saharan Africa   
141                   Senegal               Sub-Saharan Africa   
140                    Uganda               Sub-Saharan Africa   
139                   Comoros               Sub-Saharan Africa   
138       Congo (Brazzaville)               Sub-Saharan Africa   
137                      Mali               Sub-Saharan Africa   
136                    Angola               Sub-Saharan Africa   
135                     Yemen  Middle East and Northern Africa   
134                     Egypt  Middle East and Northern Africa   
133                  Bulgaria       Central and Eastern Europe   
132                  Cameroon               Sub-Saharan Africa   
131                 Sri Lanka                    Southern Asia   
130                    Malawi               Sub-Saharan Africa   
129                   Georgia       Central and Eastern Europe   
128                   Myanmar                Southeastern Asia   
127                  Botswana               Sub-Saharan Africa   
126                   Armenia       Central and Eastern Europe   
125                  Djibouti               Sub-Saharan Africa   
124                     Kenya               Sub-Saharan Africa   
123                Mauritania               Sub-Saharan Africa   
122              Sierra Leone               Sub-Saharan Africa   
121                  Ethiopia               Sub-Saharan Africa   
120                     Nepal                    Southern Asia   
119          Congo (Kinshasa)               Sub-Saharan Africa   
118                     Haiti      Latin America and Caribbean   
117                     Sudan               Sub-Saharan Africa   
116                     India                    Southern Asia   
115                   Liberia               Sub-Saharan Africa   
114                  Zimbabwe               Sub-Saharan Africa   
113                     Ghana               Sub-Saharan Africa   
112              South Africa               Sub-Saharan Africa   
111                      Iraq  Middle East and Northern Africa   
110                   Ukraine       Central and Eastern Europe   
109                      Iran  Middle East and Northern Africa   
108                Bangladesh                    Southern Asia   
107   Palestinian Territories  Middle East and Northern Africa   
106                   Tunisia  Middle East and Northern Africa   
105                Tajikistan       Central and Eastern Europe   
104                  Honduras      Latin America and Caribbean   
103                   Hungary       Central and Eastern Europe   
102                   Lebanon  Middle East and Northern Africa   
101                    Greece                   Western Europe   
100                 Swaziland               Sub-Saharan Africa   
99                   Mongolia                     Eastern Asia   
98                       Laos                Southeastern Asia

### Exercise 2 - sort by multiple columns, display the first 5 rows 
---

1. sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order 
2. display the first 5 rows of sorted data 

**Test output**:  
Records 122, 127, 147, 100, 96

In [105]:
# This sorts by Life Expectancy, then by GDP per Capita, producing the outcome above
display(happiness_df.sort_values(["Health (Life Expectancy)", "Economy (GDP per Capita)"]).head())

# This sorts by GDP per Capita, then by Life Expectancy
display(happiness_df.sort_values(["Economy (GDP per Capita)", "Health (Life Expectancy)"]).head())

Country              Region  Happiness Rank  \
122              Sierra Leone  Sub-Saharan Africa             123   
127                  Botswana  Sub-Saharan Africa             128   
147  Central African Republic  Sub-Saharan Africa             148   
100                 Swaziland  Sub-Saharan Africa             101   
96                    Lesotho  Sub-Saharan Africa              97   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
122            4.507         0.07068                   0.33024  0.95571   
127            4.332         0.04934                   0.99355  1.10464   
147            3.678         0.06112                   0.07850  0.00000   
100            4.867         0.08742                   0.71206  1.07284   
96             4.898         0.09438                   0.37545  1.04103   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
122                   0.00000  0.40840                        0.08786   
127                   0.04776  0.49495                        0.12474   
147                   0.06699  0.48879                        0.08289   
100                   0.07566  0.30658                        0.03060   
96                    0.07612  0.31767                        0.12504   

     Generosity  Dystopia Residual  
122     0.21488            2.51009  
127     0.10461            1.46181  
147     0.23835            2.72230  
100     0.18259            2.48676  
96      0.16388            2.79832

Country              Region  Happiness Rank  Happiness Score  \
119  Congo (Kinshasa)  Sub-Saharan Africa             120            4.517   
156           Burundi  Sub-Saharan Africa             157            2.905   
130            Malawi  Sub-Saharan Africa             131            4.292   
143             Niger  Sub-Saharan Africa             144            3.845   
115           Liberia  Sub-Saharan Africa             116            4.571   

     Standard Error  Economy (GDP per Capita)   Family  \
119         0.03680                   0.00000  1.00120   
156         0.08658                   0.01530  0.41587   
130         0.06130                   0.01604  0.41134   
143         0.03602                   0.06940  0.77265   
115         0.11068                   0.07120  0.78968   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
119                   0.09806  0.22605                        0.07625   
156                   0.22396  0.11850                        0.10062   
130                   0.22562  0.43054                        0.06977   
143                   0.29707  0.47692                        0.15639   
115                   0.34201  0.28531                        0.06232   

     Generosity  Dystopia Residual  
119     0.24834            2.86712  
156     0.19727            1.83302  
130     0.33128            2.80791  
143     0.19387            1.87877  
115     0.24362            2.77729

### Exercise 3 - sorting in descending order 
---
 
Sort the data by Freedom and Trust (Government Corruption) in descending order and show the Country and Region only for the last five rows

**Test output**:
136, 117, 95, 101, 111 Country and Region columns


In [106]:
display(happiness_df.sort_values(by = ["Freedom", "Trust (Government Corruption)"], ascending = False).tail())

Country                           Region  Happiness Rank  \
136                  Angola               Sub-Saharan Africa             137   
117                   Sudan               Sub-Saharan Africa             118   
95   Bosnia and Herzegovina       Central and Eastern Europe              96   
101                  Greece                   Western Europe             102   
111                    Iraq  Middle East and Northern Africa             112   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
136            4.033         0.04758                   0.75778  0.86040   
117            4.550         0.06740                   0.52107  1.01404   
95             4.949         0.06913                   0.83223  0.91916   
101            4.857         0.05062                   1.15406  0.92933   
111            4.677         0.05232                   0.98549  0.81889   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
136                   0.16683  0.10384                        0.07122   
117                   0.36878  0.10081                        0.14660   
95                    0.79081  0.09245                        0.00227   
101                   0.88213  0.07699                        0.01397   
111                   0.60237  0.00000                        0.13788   

     Generosity  Dystopia Residual  
136     0.12344            1.94939  
117     0.19062            2.20857  
95      0.24808            2.06367  
101     0.00000            1.80101  
111     0.17922            1.95335

# Cleaning the data

Data comes from a range of sources:  forms, monitoring devices, etc.  There will often be missing values, duplicate records and values that are incorrectly formatted.  These can affect summary statistics and graphs plotted from the data.

Techniques for data cleansing include:
*  removing records with missing or null data (NaN, NA, "")
*  removing duplicate rows (keeping just one, either the first or the last)

Removal of rows according to criteria, or of columns are other ways that data might be cleaned up.  


---

## Removing NaN/Dropping NA values 

pandas have functions for checking a dataframe, or column, for null values, checking a column for missing values, and functions for dropping all rows that contain null values.

* check for NA/NaN/missing values across dataframe (returns True if NA values exist)  
  `df.isnull().values.any()`  

* check for NA/NaN/missing values in specific column  
  `df["Make"].isnull().values.any()`  

* drop all rows that have NA/NaN values   
  `df.dropna()`  

* drop rows where NA/NaN values exist in specific columns  
  `df.dropna(subset = ["Make", "Model"])`  

### Exercise 4 - check for null values 
---

1. read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv 
2. check if any NA values exist in the dataframe and print the result 
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071) 



In [107]:
import pandas as pd

url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"

housing_df = pd.read_csv(url)

# Check if there are null values in the dataframe
print(housing_df.isnull().values.any())

# Display info to find columns with less that total rows
display(housing_df.info())

True
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB


None

### Exercise 5 - remove null values 
---

1. remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. remove all NA values across whole dataframe 

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

In [108]:
display(housing_df.dropna(subset = ["life_satisfaction"]))

display(housing_df.dropna())

code                      area        date  median_salary  \
613   E09000002      barking and dagenham  2011-12-01        28201.0   
614   E09000003                    barnet  2011-12-01        30237.0   
615   E09000004                    bexley  2011-12-01        28638.0   
616   E09000005                     brent  2011-12-01        26772.0   
617   E09000006                   bromley  2011-12-01        28163.0   
618   E09000007                    camden  2011-12-01        35967.0   
619   E09000008                   croydon  2011-12-01        28979.0   
620   E09000009                    ealing  2011-12-01        28646.0   
621   E09000010                   enfield  2011-12-01        26412.0   
622   E09000011                 greenwich  2011-12-01        29932.0   
623   E09000012                   hackney  2011-12-01        33512.0   
624   E09000013    hammersmith and fulham  2011-12-01        36143.0   
625   E09000014                  haringey  2011-12-01        28032.0   
626   E09000015                    harrow  2011-12-01        27540.0   
627   E09000016                  havering  2011-12-01        28415.0   
628   E09000017                hillingdon  2011-12-01        33589.0   
629   E09000018                  hounslow  2011-12-01        32759.0   
630   E09000019                 islington  2011-12-01        35996.0   
631   E09000020    kensington and chelsea  2011-12-01        31491.0   
632   E09000021      kingston upon thames  2011-12-01        28997.0   
633   E09000022                   lambeth  2011-12-01        34836.0   
634   E09000023                  lewisham  2011-12-01        29567.0   
635   E09000024                    merton  2011-12-01        28066.0   
636   E09000025                    newham  2011-12-01        29998.0   
637   E09000026                 redbridge  2011-12-01        29829.0   
638   E09000027      richmond upon thames  2011-12-01        32197.0   
639   E09000028                 southwark  2011-12-01        36124.0   
640   E09000029                    sutton  2011-12-01        26743.0   
641   E09000030             tower hamlets  2011-12-01        45970.0   
642   E09000031            waltham forest  2011-12-01        28005.0   
643   E09000032                wandsworth  2011-12-01        32033.0   
644   E09000033               westminster  2011-12-01        37406.0   
645   E12000001                north east  2011-12-01        23250.0   
646   E12000002                north west  2011-12-01        24167.0   
647   E12000003  yorkshire and the humber  2011-12-01        24119.0   
648   E12000004             east midlands  2011-12-01        24002.0   
649   E12000005             west midlands  2011-12-01        24291.0   
651   E12000007                    london  2011-12-01        34396.0   
652   E12000008                south east  2011-12-01        27881.0   
653   E12000009                south west  2011-12-01        24306.0   
656   E92000001                   england  2011-12-01        26488.0   
660   N92000002          northern ireland  2011-12-01        23034.0   
661   S92000003                  scotland  2011-12-01        25291.0   
662   W92000004                     wales  2011-12-01        23397.0   
664   E09000002      barking and dagenham  2012-12-01        33131.0   
665   E09000003                    barnet  2012-12-01        29594.0   
666   E09000004                    bexley  2012-12-01        27440.0   
667   E09000005                     brent  2012-12-01        27577.0   
668   E09000006                   bromley  2012-12-01        28082.0   
669   E09000007                    camden  2012-12-01        36962.0   
670   E09000008                   croydon  2012-12-01        28974.0   
671   E09000009                    ealing  2012-12-01        31104.0   
672   E09000010                   enfield  2012-12-01        27123.0   
673   E09000011                 greenwich  2012-12-01        30097.0   
674   E09000012                   hackney  2012-12-

code                    area        date  median_salary  \
613   E09000002    barking and dagenham  2011-12-01        28201.0   
614   E09000003                  barnet  2011-12-01        30237.0   
615   E09000004                  bexley  2011-12-01        28638.0   
616   E09000005                   brent  2011-12-01        26772.0   
617   E09000006                 bromley  2011-12-01        28163.0   
618   E09000007                  camden  2011-12-01        35967.0   
619   E09000008                 croydon  2011-12-01        28979.0   
620   E09000009                  ealing  2011-12-01        28646.0   
621   E09000010                 enfield  2011-12-01        26412.0   
622   E09000011               greenwich  2011-12-01        29932.0   
623   E09000012                 hackney  2011-12-01        33512.0   
624   E09000013  hammersmith and fulham  2011-12-01        36143.0   
625   E09000014                haringey  2011-12-01        28032.0   
626   E09000015                  harrow  2011-12-01        27540.0   
627   E09000016                havering  2011-12-01        28415.0   
628   E09000017              hillingdon  2011-12-01        33589.0   
629   E09000018                hounslow  2011-12-01        32759.0   
630   E09000019               islington  2011-12-01        35996.0   
631   E09000020  kensington and chelsea  2011-12-01        31491.0   
632   E09000021    kingston upon thames  2011-12-01        28997.0   
633   E09000022                 lambeth  2011-12-01        34836.0   
634   E09000023                lewisham  2011-12-01        29567.0   
635   E09000024                  merton  2011-12-01        28066.0   
636   E09000025                  newham  2011-12-01        29998.0   
637   E09000026               redbridge  2011-12-01        29829.0   
638   E09000027    richmond upon thames  2011-12-01        32197.0   
639   E09000028               southwark  2011-12-01        36124.0   
640   E09000029                  sutton  2011-12-01        26743.0   
641   E09000030           tower hamlets  2011-12-01        45970.0   
642   E09000031          waltham forest  2011-12-01        28005.0   
643   E09000032              wandsworth  2011-12-01        32033.0   
644   E09000033             westminster  2011-12-01        37406.0   
651   E12000007                  london  2011-12-01        34396.0   
656   E92000001                 england  2011-12-01        26488.0   
664   E09000002    barking and dagenham  2012-12-01        33131.0   
665   E09000003                  barnet  2012-12-01        29594.0   
666   E09000004                  bexley  2012-12-01        27440.0   
667   E09000005                   brent  2012-12-01        27577.0   
668   E09000006                 bromley  2012-12-01        28082.0   
669   E09000007                  camden  2012-12-01        36962.0   
670   E09000008                 croydon  2012-12-01        28974.0   
671   E09000009                  ealing  2012-12-01        31104.0   
672   E09000010                 enfield  2012-12-01        27123.0   
673   E09000011               greenwich  2012-12-01        30097.0   
674   E09000012                 hackney  2012-12-01        34803.0   
675   E09000013  hammersmith and fulham  2012-12-01        35136.0   
676   E09000014                haringey  2012-12-01        29116.0   
677   E09000015                  harrow  2012-12-01        26738.0   
678   E09000016                havering  2012-12-01        28326.0   
679   E09000017              hillingdon  2012-12-01        33822.0   
680   E09000018                hounslow  2012-12-01        31821.0   
681   E09000019               islington  2012-12-01        36653.0   
682   E09000020  kensington and chelsea  2012-12-01        30726.0   
683   E09000021    kingston upon thames  2012-12-01        29067.0   
684   E09000022                 lambeth  2012-12-01        35280.0   
685   E09000023                lewisham  2012-12-01        29995.0   
686   E09000024      

## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])` 

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 6 - Removing duplicate entries 
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01* 

In [109]:
display(housing_df.drop_duplicates(subset = ["area"]))

code                      area        date  median_salary  \
0   E09000001            city of london  1999-12-01        33020.0   
1   E09000002      barking and dagenham  1999-12-01        21480.0   
2   E09000003                    barnet  1999-12-01        19568.0   
3   E09000004                    bexley  1999-12-01        18621.0   
4   E09000005                     brent  1999-12-01        18532.0   
5   E09000006                   bromley  1999-12-01        16720.0   
6   E09000007                    camden  1999-12-01        23677.0   
7   E09000008                   croydon  1999-12-01        19563.0   
8   E09000009                    ealing  1999-12-01        20580.0   
9   E09000010                   enfield  1999-12-01        19289.0   
10  E09000011                 greenwich  1999-12-01        21236.0   
11  E09000012                   hackney  1999-12-01        23249.0   
12  E09000013    hammersmith and fulham  1999-12-01        25000.0   
13  E09000014                  haringey  1999-12-01        18783.0   
14  E09000015                    harrow  1999-12-01        20596.0   
15  E09000016                  havering  1999-12-01        17165.0   
16  E09000017                hillingdon  1999-12-01        24002.0   
17  E09000018                  hounslow  1999-12-01        20155.0   
18  E09000019                 islington  1999-12-01        25113.0   
19  E09000020    kensington and chelsea  1999-12-01        20646.0   
20  E09000021      kingston upon thames  1999-12-01        19302.0   
21  E09000022                   lambeth  1999-12-01        23151.0   
22  E09000023                  lewisham  1999-12-01        20580.0   
23  E09000024                    merton  1999-12-01        18962.0   
24  E09000025                    newham  1999-12-01        18862.0   
25  E09000026                 redbridge  1999-12-01        19580.0   
26  E09000027      richmond upon thames  1999-12-01        22321.0   
27  E09000028                 southwark  1999-12-01        22784.0   
28  E09000029                    sutton  1999-12-01        19582.0   
29  E09000030             tower hamlets  1999-12-01        26376.0   
30  E09000031            waltham forest  1999-12-01        18547.0   
31  E09000032                wandsworth  1999-12-01        21321.0   
32  E09000033               westminster  1999-12-01        24447.0   
33  E12000001                north east  1999-12-01        16282.0   
34  E12000002                north west  1999-12-01        16977.0   
35  E12000003  yorkshire and the humber  1999-12-01        16527.0   
36  E12000004             east midlands  1999-12-01        16392.0   
37  E12000005             west midlands  1999-12-01        17000.0   
38  E12000006                      east  1999-12-01        18000.0   
39  E12000007                    london  1999-12-01        22487.0   
40  E12000008                south east  1999-12-01        18737.0   
41  E12000009                south west  1999-12-01        16727.0   
42  E13000001              inner london  1999-12-01            NaN   
43  E13000002              outer london  1999-12-01            NaN   
44  E92000001                   england  1999-12-01        17939.0   
45  K02000001            united kingdom  1999-12-01        17803.0   
46  K03000001             great britain  1999-12-01        17866.0   
47  K04000001         england and wales  1999-12-01        17974.0   
48  N92000002          northern ireland  1999-12-01        15798.0   
49  S92000003                  scotland  1999-12-01        16914.0   
50  W92000004                     wales  1999-12-01        16457.0   

    life_satisfaction mean_salary recycling_pct  population_size  \
0                 NaN       48922             0           6581.0   
1                 NaN       23620             3         162444.0   
2                 NaN       23128             8         313469.0   
3                 NaN       21386            18         217458.0   
4                 NaN       

# Normalising Data  
When we normalise data, we remodel a numeric column in a dataframe to be on a standard scale (e.g. 0 or 1).   

For example if we had a column of BMI scores, we could normalise that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

To normalise we need to:
*   write a function, with the dataframe as a parameter, which will look at each row in dataframe column and return either a value in the normalised scale (e.g. 0,1 or 1,2,3,4) depending on that value.

For example:  
```
def normalise_bmi(df):
  if df['bmi'] >= 25:
    return 1
  else:
    return 0

df["bmi"] = df.apply(normalise_bmi, axis=1)
```
This code reassigns the values in the column "bmi" by sending each row one after the other to the normalise_bmi function, which will check the value in the "bmi" column and return either 0 or 1 depending on the value in the "bmi" column.

### Exercise 7 - normalise data set
---

Create a function called **normalise_income(df)** that will return the values 1, 2 or 3 to represent low income, middle income and high income.  If the value in `df['median_salary']` is less than 27441 (the median), return 1, otherwise if it is less than 30932 (the upper quartile) return 2 and otherwise return 3.

Apply the normalise_income(df) function to the `median_salary` column.

*NOTE:  this operation will change the original dataframe so if you run it twice, everything in the median_salary column will change to 1 (as it had already been reduced to 1, 2 or 3 - if this happens, run the code in Exercise 4 again to get the original data again from the file.*

**Test output**:  
The maximum value of the column df['median_salary'] will be 3 and the minimum value will be 1  

In [110]:
def normalise_income(df):
    if df['median_salary'] < 27441:
        return 1
    elif df['median_salary'] < 30932:
        return 2
    else:
        return 3

housing_df["median_salary"] = housing_df.apply(normalise_income, axis=1)

# Test
display(housing_df["median_salary"].max())
display(housing_df["median_salary"].min())
    

3

1

### Exercise 8 - normalise the number of jobs column
---

Using what you have learnt from Exercise 7:  
*  use `df.describe()` to find the median, upper quartile and maximum for the number_of_jobs column  
*  create a function called **normalise_jobs(df)** that will return 1 if the `number_of_jobs` is below the median, 2 if the `number_of_jobs` is below the upper quartile or 3 otherwise.
*  normalise the `number_of_jobs` column by applying the function `normalise_jobs`.

**Test output**:  
The maximum value of the column df['number_of_jobs'] will be 3 and the minimum value will be 1  

In [111]:
pd.set_option('display.max_rows', 100)

display(housing_df.describe())

def normalise_jobs(df):
    if df["number_of_jobs"] < 157000:
        return 1
    elif df["number_of_jobs"] < 2217000:
        return 2
    else:
        return 3

housing_df["number_of_jobs"] = housing_df.apply(normalise_jobs, axis=1)

# Test
display(housing_df["number_of_jobs"].max())
display(housing_df["number_of_jobs"].min())

median_salary  life_satisfaction  population_size  number_of_jobs  \
count    1071.000000         352.000000     1.018000e+03    9.310000e+02   
mean        1.776844           7.485057     6.042576e+06    3.188095e+06   
std         0.840375           0.198451     1.526810e+07    8.058302e+06   
min         1.000000           7.000000     6.581000e+03    4.700000e+04   
25%         1.000000           7.350000     2.243458e+05    9.450000e+04   
50%         2.000000           7.510000     2.946035e+05    1.570000e+05   
75%         3.000000           7.640000     4.630098e+06    2.217000e+06   
max         3.000000           7.960000     6.643555e+07    3.575000e+07   

          area_size  no_of_houses  borough_flag  
count  6.660000e+02  6.660000e+02   1071.000000  
mean   3.724903e+05  8.814682e+05      0.647059  
std    2.157060e+06  3.690376e+06      0.478108  
min    3.150000e+02  5.009000e+03      0.000000  
25%    2.960000e+03  8.763550e+04      0.000000  
50%    4.323000e+03  1.024020e+05      1.000000  
75%    8.220000e+03  1.262760e+05      1.000000  
max    1.330373e+07  2.417217e+07      1.000000

3

1

## Exercise 9 - normalise into a new column
---

Create a new function and code to normalise the `no_of_houses` column BUT this time, instead of assigning the result to `df['no_of_houses']` assign it to a new column called `df['housing_volume']`

**Test output**:  
The maximum value of the column df['housing_volume'] will be 3 and the minimum value will be 1 

In [112]:
def normalise_no_of_houses(df):
    if df["no_of_houses"] < 102402:
        return 1
    elif df["no_of_houses"] < 126276:
        return 2
    else:
        return 3

housing_df["housing_volume"] = housing_df.apply(normalise_no_of_houses, axis=1)

# Test
display(housing_df["housing_volume"].max())
display(housing_df["housing_volume"].min())

3

1

### Exercise 10 - normalise boroughs
---

Normalise the `area_size` column so that all values below mean are represented as 0 and otherwise are 1.  Assign the output to a new column called `area_size_normalised`.  

**Test output**:  
`area_size_normalised` column will contain both 0s and 1s.  The position of the first row with value 1 will be 0 and the position of the first row with value 0 will be 102.


In [114]:
def normalise_area_size(df):
    if df["area_size"] < housing_df['area_size'].mean():
        return 0
    else:
        return 1

housing_df["area_size_normalised"] = housing_df.apply(normalise_area_size, axis=1)

# Test
pd.options.display.max_rows = 999
housing_df['area_size_normalised'][0:103]

0      1
1      1
2      1
3      1
4      1
5      1
6      1
7      1
8      1
9      1
10     1
11     1
12     1
13     1
14     1
15     1
16     1
17     1
18     1
19     1
20     1
21     1
22     1
23     1
24     1
25     1
26     1
27     1
28     1
29     1
30     1
31     1
32     1
33     1
34     1
35     1
36     1
37     1
38     1
39     1
40     1
41     1
42     1
43     1
44     1
45     1
46     1
47     1
48     1
49     1
50     1
51     1
52     1
53     1
54     1
55     1
56     1
57     1
58     1
59     1
60     1
61     1
62     1
63     1
64     1
65     1
66     1
67     1
68     1
69     1
70     1
71     1
72     1
73     1
74     1
75     1
76     1
77     1
78     1
79     1
80     1
81     1
82     1
83     1
84     1
85     1
86     1
87     1
88     1
89     1
90     1
91     1
92     1
93     1
94     1
95     1
96     1
97     1
98     1
99     1
100    1
101    1
102    0
Name: area_size_normalised, dtype: int64

# Reflection
----

## What skills have you demonstrated in completing this notebook?

Your answer: Ability to sort, clean and normalise data using pandas library functions. 

## What caused you the most difficulty?

Your answer: No difficulties.